rest-service/manager_rest/rest/resources_v3_1/searches.py

```diff
@@ -551,3 +554,21 @@ def scaling_group_name_matches(scaling_group_name, constraints, search_value):
         return scaling_group_name == search_value
 
     return True
+
+
+def extend_node_type_valid_values(deployment_id, valid_values):
+    """Extend the list of valid node types based on the type hierarchy.
+    Include also ancestors of the types listed as valid_values."""
+    sm = get_storage_manager()
+    nodes = sm.list(
+        models.Node,
+        filters={'deployment_id': deployment_id},
+        include='id,type,type_hierarchy',
+        get_all_results=True)
+    results = []
+    for value in valid_values:
+        for node in nodes:
+            if value in node.type_hierarchy:
+                th = node.type_hierarchy
+                results.extend(th[th.index(value):])
```

Comment:
Luckily `type_hierarchy` will preserve the order of inheritance with the root node being the first in the list and the actual node.type - the last.  For example:

```json
"id": "myserver",
"type": "webserver",
"type_hierarchy": [
    "cloudify.nodes.Root",
    "cloudify.nodes.SoftwareComponent",
    "cloudify.nodes.WebServer",
    "webserver"
]
```